src/tools/wasm-reduce.cpp

```diff
@@ -1139,33 +1142,43 @@ struct Reducer
     return false;
   }
 
-  // try to replace a concrete value with a trivial constant
+  // Try to replace a concrete value with a trivial constant.
   bool tryToReduceCurrentToConst() {
     auto* curr = getCurrent();
-    if (curr->is<Const>()) {
-      return false;
-    }
-    // try to replace with a trivial value
-    if (curr->type.isNullable()) {
+
+    // References.
+    if (curr->type.isNullable() && !curr->is<RefNull>()) {
       RefNull* n = builder->makeRefNull(curr->type.getHeapType());
       return tryToReplaceCurrent(n);
     }
+
+    // Tuples.
     if (curr->type.isTuple() && curr->type.isDefaultable()) {
       Expression* n =
         builder->makeConstantExpression(Literal::makeZeros(curr->type));
+      if (ExpressionAnalyzer::equal(n, curr)) {
+        return false;
+      }
       return tryToReplaceCurrent(n);
     }
+
+    // Numbers. We try to replace them with a 0 or a 1.
     if (!curr->type.isNumber()) {
       return false;
     }
-    // It's a number: try to replace it with a 0 or a 1 (trying more values
-    // could make sense too, but these handle most cases).
     auto* c = builder->makeConst(Literal::makeZero(curr->type));
+    if (ExpressionAnalyzer::equal(c, curr)) {
```

Comment:
Can we use `Literal::isZero` or something for checking before making? (It doesn't seem to have `Literal::isOne` though.. Can we add it?)

Reply:
Good idea, yeah, we can be a little more efficient here. I did that now.

Separately, not sure if it's worth adding `Literall::isOne`. It would save a little code but it is rare to check that I suspect...

Reply:
Sounds good